### PR TITLE
Revert "Revert "resolve: Avoid "self-confirming" import resolutions in one more case""

### LINKED ISSUE #77586: `use dep1::foo as dep1` is considered ambiguous
`src/lib.rs`:
```
pub mod foo {
    pub use dep1::foo as dep1;
}
```

`dep1/src/lib.rs`:
```
pub mod foo {}
```

I expected to see this happen: build succeeds.

Instead, this happened:
```
error[E0432]: unresolved import `dep1::foo`
 --> src/lib.rs:2:13
  |
2 |     pub use dep1::foo as dep1;
  |             ^^^^^^^^^^^^^^^^^ no `foo` in `foo`

error[E0659]: `dep1` is ambiguous (name vs any other name during import resolution)
 --> src/lib.rs:2:13
  |
2 |     pub use dep1::foo as dep1;
  |             ^^^^ ambiguous name
  |
  = note: `dep1` could refer to a crate passed with `--extern`
  = help: use `::dep1` to refer to this crate unambiguously
note: `dep1` could also refer to the module imported here
 --> src/lib.rs:2:13
  |
2 |     pub use dep1::foo as dep1;
  |             ^^^^^^^^^^^^^^^^^
  = help: use `self::dep1` to refer to this module unambiguously

error: aborting due to 2 previous errors

Some errors have detailed explanations: E0432, E0659.
For more information about an error, try `rustc --explain E0432`.
```

### Meta

working beta: `rustc --version --verbose`:
```
rustc 1.47.0-beta.9 (d6646f647 2020-10-03)
binary: rustc
commit-hash: d6646f64790018719caebeafd352a92adfa1d75a
commit-date: 2020-10-03
host: x86_64-unknown-linux-gnu
release: 1.47.0-beta.9
LLVM version: 11.0
```

non-working nightly: `rustc +nightly --version --verbose`
```
rustc 1.49.0-nightly (beb5ae474 2020-10-04)
binary: rustc
commit-hash: beb5ae474d2835962ebdf7416bd1c9ad864fe101
commit-date: 2020-10-04
host: x86_64-unknown-linux-gnu
release: 1.49.0-nightly
LLVM version: 11.0
```

This broke [`rust-systemd`](https://crates.io/crates/systemd)'s build (https://travis-ci.com/github/jmesmon/rust-systemd/jobs/395293939) (broken versions are `v0.6.0` and `v0.5.0` which use `edition = "2018"`).

I'm not sure if there is a simpler single-crate break (there could be).

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
Specifically, this reverts commit b20bce8ce54ea9d47c2e3eb0b17cbb6baf916ae2 from #77421 to fix #77586.

The lang team has decided that for the time being we want to avoid the breakage here (perhaps for a future edition; though almost certainly not the upcoming one), though a future PR may want to add a lint around this case (and perhaps others) which are unlikely to be readable code.

r? @petrochenkov to confirm this is the right way to fix #77586.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
beta-nominating since this targets a fix to a beta regression, #77586.

**Comment by petrochenkov** [comment]
It would be nice to keep the test, but otherwise r=me.

File: src/test/ui/imports/issue-62767.rs [hunk 1]
```diff
@@ -20,9 +22,9 @@ mod foo {
     }
 }
 
+use bar::bar;
 use foo::*;
-use bar::bar; //~ ERROR unresolved import `bar::bar`
-              //~| ERROR inconsistent resolution for an import
```
Comment (petrochenkov) [inline src/test/ui/imports/issue-62767.rs:25]: IIRC, the import order was important for reproducing the ICE.

Comment (Mark-Simulacrum) [reply src/test/ui/imports/issue-62767.rs:25]: Good catch, I was editing in a different directory and didn't notice rustfmt reformatting the file. Adjusted back.

Comment (pnkfelix) [reply src/test/ui/imports/issue-62767.rs:25]: (Is there a way to tell rustfmt not to do this for this file, via an embedded comment or attribute in the source?)

**Comment by spastorino** [comment]
[discussed](https://zulip-archive.rust-lang.org/238009tcompilermeetings/78755weeklymeeting2020111254818.html#216480984) in T-compiler meeting.

@rustbot modify labels: beta-accepted

File: src/test/ui/imports/issue-62767.rs (hunk 1 above)
Comment (Mark-Simulacrum) [reply src/test/ui/imports/issue-62767.rs:25]: It's not going to do it via x.py fmt, but I was editing in a different workspace and hadn't copied my vim config to it.

**Pull request merged**